crates/services/relayer/src/service.rs

```diff
@@ -207,11 +211,54 @@ impl SharedState {
     /// some period of time.
     pub async fn await_synced(&self) -> anyhow::Result<()> {
         let mut rx = self.synced.clone();
-        if !rx.borrow_and_update().deref() {
+        if rx.borrow_and_update().deref().is_none() {
+            rx.changed().await?;
+        }
+        Ok(())
+    }
+
+    /// Wait until at least the given height is synced.
+    pub async fn await_at_least_synced(
+        &self,
+        height: &DaBlockHeight,
+    ) -> anyhow::Result<()> {
+        let mut rx = self.synced.clone();
+        while rx
+            .borrow_and_update()
+            .deref()
+            .map_or(true, |h| h >= *height)
+        {
             rx.changed().await?;
         }
         Ok(())
     }
+
+    /// Get a message if it has been synced
+    /// and is <= the given height.
+    pub fn get_message(
+        &self,
+        id: &MessageId,
+        da_height: &DaBlockHeight,
+    ) -> anyhow::Result<Option<CompressedMessage>>
+    where
+        D: StorageInspect<Messages, Error = fuel_core_storage::Error>,
+    {
+        Ok(self
```

Comment:
Maybe you want to put it also into the `RelayerDB` trait?

Reply:
I don't see why? It's only used here.